### PR TITLE
Issue #998: Clarify layouts terminology

### PR DESCRIPTION
**Pull request by docwilmot**
Fixes https://github.com/backdrop/backdrop-issues/issues/998

Replaces https://github.com/backdrop/backdrop/pull/984

### CODE REVIEW

File: core/modules/layout/layout.install [hunk 1]
```diff
@@ -34,3 +34,21 @@ function layout_disable() {
...
+  $layouts= config_get_names_with_prefix('layout');
+  foreach ($layouts as $layout_name) {
+    $config = config($layout_name);
+    $configs = $config->get();
+    if (isset($configs['layout'])) {
+      $configs['layout_template'] = $configs['layout'];
+      unset($configs['layout']);
+    }
+  $config->setData($configs);
+  $config->save();
```
Comment (quicksketch) [inline core/modules/layout/layout.install]: Lines 51 and 52 should be indented 2 additional spaces.

File: core/modules/layout/layout.module [hunk 2]
```diff
@@ -1139,6 +1139,28 @@ function layout_get_layout_info($layout_name = NULL) {
...
+ *   The name of the layout template, e.g. "one_column" or
+ *   "three_three_four_column". If no layout template is specified, all layout 
+ *  info will be returned.
+ * @return array|boolean
+ *   The layout template information, as returned by either a stand-alone 
+ *   template .info file, or through a module's hook_layout_info().
+ *
+ * @see layout_get_layout_template_info()
+ *
+ * @deprecated; replaced by layout_get_layout_template_info().
```
Comment (quicksketch) [inline core/modules/layout/layout.module]: :+1: Thanks for this wrapper. Definitely important for compatibility.

Let's use the format `@deprecated since 1.4.0` here.

File: core/modules/simpletest/tests/upgrade/upgrade.test [hunk 3]
```diff
@@ -207,7 +207,15 @@ abstract class UpgradePathTestCase extends BackdropWebTestCase {
...
-    if (!$this->assertText(t('No pending updates.'), 'No pending updates at the end of the update process.')) {
+
+    // Install new updates if any.
+    if ($this->assertText(t('Updates have been found that need to be applied'))) {
+      $this->backdropPost(NULL, array(), t('Apply pending updates'));
+      if (!$this->assertText(t('Updates were attempted'))) {
+        return FALSE;
+      }
+    }
+    elseif (!$this->assertText(t('No pending updates.'), 'No pending updates at the end of the update process.')) {
```
Comment (quicksketch) [inline core/modules/simpletest/tests/upgrade/upgrade.test]: Hm, is this an unrelated change? Maybe move to a separate issue?

Comment (docwilmot) [inline core/modules/simpletest/tests/upgrade/upgrade.test]: I wondered if this required a separate issue in fact, but it was necessary for tests to pass. 

The current `upgrade.test` `performUpgrade()` runs on a database without `layout` module, so `layout_update_N()` doesnt run and the function expects that after a D7 upgrade there would be 'no pending updates'. So when layout installs with my new `layout_update_1000(),` `performUpgrade()` finds this and terminates the upgrade. This little code block anticipates this.

An alternative is to not call `performUpgrade(),` and to duplicate it in `layout.test` instead, but it seemed wiser to fix the `upgrade.test` version instead. 

What else could we do here?

**Comment by klonos** [comment]
The label in the "Add layout" form should be "Layout template" (singular - it now has a trailing "s").

File: core/modules/layout/layout.admin.inc [hunk 4]
```diff
@@ -221,18 +221,18 @@ function layout_settings_form($form, &$form_state, Layout $layout) {
     '#required' => TRUE,
     '#access' => !$layout->isDefault(),
   );
-  $form['layout'] = array(
-    '#title' => t('Layout'),
+  $form['layout_template'] = array(
+    '#title' => t('Layout templates'),
```
Comment (jenlampton) [inline core/modules/layout/layout.admin.inc]: This should be `Layout template` as only one template can be selected for any given `Layout`.

File: core/modules/update/update.module [hunk 5]
```diff
@@ -193,10 +193,10 @@ function update_menu() {
   }
   // Customize the titles of the action links depending on where they appear.
   // We use += array() to let the translation extractor find these menu titles.
-  $items['admin/reports/updates/install'] += array('title' => 'Install module, theme or layout');
+  $items['admin/reports/updates/install'] += array('title' => 'Install module, theme, or layout template');
   $items['admin/modules/install'] += array('title' => 'Install module');
   $items['admin/appearance/install'] += array('title' => 'Install theme');
-  $items['admin/structure/layouts/install'] += array('title' => 'Install layout');
+  $items['admin/structure/layouts/install'] += array('title' => 'Install layout template');
```
Comment (jenlampton) [inline core/modules/update/update.module]: Let's hold off on this change for now, as we should update the links for modules & themes at the same time. Perhaps as part of the addition of project browser!  See https://github.com/backdrop/backdrop-issues/issues/1757

File: core/modules/update/update.module [hunk 6]
```diff
@@ -193,10 +193,10 @@ function update_menu() {
   }
   // Customize the titles of the action links depending on where they appear.
   // We use += array() to let the translation extractor find these menu titles.
-  $items['admin/reports/updates/install'] += array('title' => 'Install module, theme or layout');
+  $items['admin/reports/updates/install'] += array('title' => 'Install module, theme, or layout template');
```
Comment (jenlampton) [inline core/modules/update/update.module]: Let's hold off on this change for now, as we should update the links for modules & themes at the same time. Perhaps as part of the addition of project browser! See backdrop/backdrop-issues#1757

**Comment by docwilmot** [comment]
@jenlampton thanks for the review. Just one thing left to RTBC this, is a decision on that little code block in https://github.com/backdrop/backdrop/pull/1307#discussion-diff-57506801 which @quicksketch  commented on. Should I leave this here in this PR, or make a separate PR. As it stands, it is not going  to bother any other test, but I suppose it is a bugfix of a sort. What you think?

New issue done at https://github.com/backdrop/backdrop-issues/issues/1759 FTR.

File: core/modules/simpletest/tests/upgrade/upgrade.test (hunk 3 above)
Comment (quicksketch) [inline core/modules/simpletest/tests/upgrade/upgrade.test]: >  So when layout installs with my new layout_update_1000(), performUpgrade() finds this and terminates the upgrade. This little code block anticipates this.

Ah right. So this code block adjusts how the upgrade process works (both for tests and end-users). This means that you'd need to run update.php twice in order to upgrade a site from D7 to Backdrop. In other situations where we've needed to run updates to modules that did not exist in Drupal 7 but are enabled by default in Backdrop, we put the updates in system.install. I think that would be appropriate for this situation as well, instead of making `layout_update_1000()`.

**Comment by docwilmot** [comment]
OK new PR force pushed with all recommendations. This should be RTBC-able.

**Comment by jenlampton** [comment]
yep, looks great!

**Comment by quicksketch** [comment]
I rerolled this in https://github.com/backdrop/backdrop/pull/1382.
